server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

```diff
@@ -145,72 +151,88 @@ private PublishWithJoinResponse handleIncomingPublishRequest(BytesTransportReque
-                        final CoordinationMetadata.VotingConfiguration lastCommittedConfiguration;
-
-                        // Close stream early to release resources used by the de-compression as early as possible
-                        try (StreamInput input = in) {
-                            diff = ClusterState.readDiffFrom(input, lastSeen.nodes().getLocalNode());
-                            if (includesLastCommittedData) {
-                                clusterUuidCommitted = in.readBoolean();
-                                lastCommittedConfiguration = new CoordinationMetadata.VotingConfiguration(in);
-                            } else {
-                                clusterUuidCommitted = false;
-                                lastCommittedConfiguration = null;
-                            }
-                            assert input.read() == -1;
-                        }
-                        incomingState = diff.apply(lastSeen); // might throw IncompatibleClusterStateVersionException
-                        if (includesLastCommittedData) {
-                            final var adjustedMetadata = incomingState.metadata()
-                                .withLastCommittedValues(clusterUuidCommitted, lastCommittedConfiguration);
-                            if (adjustedMetadata != incomingState.metadata()) {
-                                incomingState = ClusterState.builder(incomingState).metadata(adjustedMetadata).build();
-                            }
-                        }
-                    } catch (IncompatibleClusterStateVersionException e) {
-                        incompatibleClusterStateDiffReceivedCount.incrementAndGet();
-                        throw e;
-                    } catch (Exception e) {
-                        logger.warn("unexpected error while deserializing an incoming cluster state", e);
-                        assert false : e;
-                        throw e;
-                    }
+                    final ClusterState incomingState = deserializeAndApplyDiff(request, in, lastSeen);
                     compatibleClusterStateDiffReceivedCount.incrementAndGet();
                     logger.debug(
                         "received diff cluster state version [{}] with uuid [{}], diff size [{}]",
                         incomingState.version(),
                         incomingState.stateUUID(),
                         request.bytes().length()
                     );
-                    final PublishWithJoinResponse response = acceptState(incomingState);
-                    lastSeenClusterState.compareAndSet(lastSeen, incomingState);
-                    return response;
+                    acceptState(incomingState, publishResponseListener.delegateFailure((delegate, response) -> {
+                        lastSeenClusterState.compareAndSet(lastSeen, incomingState);
+                        delegate.onResponse(response);
+                    }));
```

Comment:
Sorry one more tiny thing, I would write this sort of thing using `ActionListener#map` to avoid any doubt that we complete the listener on all paths:

```suggestion
                    acceptState(incomingState, publishResponseListener.map(response -> {
                        lastSeenClusterState.compareAndSet(lastSeen, incomingState);
                        return response;
                    }));
```

Not a huge deal here ofc but a useful trick in more complex cases for sure.

Reply:
Hehe yeah it took me a while to work out which helper to use here. Will make the change.